client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/CRUDDocumentationIT.java

```diff
@@ -240,9 +225,34 @@ public void onFailure(Exception e) {
             }
             // end::index-optype
         }
+        {
+            IndexRequest request = new IndexRequest("posts", "doc", "async").source("field", "value");
+
+            // tag::index-execute-async
+            client.indexAsync(request, new ActionListener<IndexResponse>() {
+                @Override
+                public void onResponse(IndexResponse indexResponse) {
+                    // <1>
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    // <2>
+                }
+            });
+            // end::index-execute-async
+
+            assertTrue(awaitBusy(() -> {
+                try {
+                    return client.exists(new GetRequest("posts", "doc", "async"));
+                } catch (IOException e) {
+                    return false;
+                }
+            }));
```

Comment:
what happens if we don't wait for this one?

Reply:
I was also thinking, given that this is a recurring problem and we don't want to have latches to show in our docs, shall we make this more clear doing something through a listener that blocks (without showing in the snippets what the listener really does)?

Reply:
> what happens if we don't wait for this one?

After each test method, the `ESRestTestCase.cleanUpCluster()` is executed. As you know this method deletes the cluster indices, waits for the pending cluster states tasks to finish and logs a message if there is still inflight requests. 

By not waiting for the async request to respond, it's possible that an index got re created because the index (or update) operation is processed after the indices have been deleted and s recreates the index and breaks following tests. 

I'm pretty sure I saw this error, but sadly I can't find a test execution log that clearly shows this behavior. In any way I think it's better to wait for the completion of any write operation that could recreate an index for the reason I mentioned.

Reply:
makes sense, thanks for explaining.

Reply:
what do you think about the listener idea above?

Reply:
I'd prefer to have a blocking listener than assertBusy loops but I guess we'll have to declare a simple listener just for the doc purpose like:

```
//tag::listener
ActionListener<IndexResponse> listener = new ActionListener<IndexResponse>() {
                @Override
                public void onResponse(IndexResponse indexResponse) {
                    // <1>
                }

                @Override
                public void onFailure(Exception e) {
                    // <2>
                }
            };
//end::listener
```

Then, out of documented snippet replace the listener by a blocking one:

```
// Replace the listener by a blocking one
final CountDownLatch latch = new CountDownLatch(1);
listener = new ActionListener<IndexResponse> {
.. use latch in on response/onfailure methods here
}
```

And finally show the async execution snippet using the listener reference:
```
// tag::index-execute-async
client.indexAsync(request, listener);
//end::async
```

Is that what you have in mind?

Reply:
yea this sounds good to me.

Reply:
This will have an impact on documentation (2 code snippets instead of one), should we align all documentation for async execution to be like this? I think so.

Reply:
yes I would adapt the docs, let me know if you prefer me to take this over given that it's becoming a bigger task.

Reply:
I can do it. Do you prefer a follow up PR or be part of this one?

Reply:
I don't mind either way